packages/babel-plugin-proposal-destructuring-private/src/index.ts

```diff
@@ -0,0 +1,199 @@
+        .unshiftContainer(
+          "body",
+          variableDeclaration("let", [variableDeclarator(node.param, ref)]),
+        );
+      node.param = cloneNode(ref);
+      // the pattern will be handled by VariableDeclaration visitor.
+    },
+    ForXStatement(path) {
+      const { node, scope } = path;
+      const leftPath = path.get("left");
+      if (leftPath.isVariableDeclaration()) {
+        const left = leftPath.node;
+        if (!hasPrivateKeys(left.declarations[0].id)) return;
+        // for (const { #x: x } of cls) body;
+        // transforms to:
+        // for (const ref of cls) { const { #x: x } = ref; body; }
+        const temp = scope.generateUidIdentifier("ref");
+        node.left = variableDeclaration(left.kind, [
+          variableDeclarator(temp, null),
+        ]);
+        left.declarations[0].init = cloneNode(temp);
+        unshiftForXStatementBody(path, [left]);
+        scope.crawl();
+        // the pattern will be handled by VariableDeclaration visitor.
+      } else if (leftPath.isPattern()) {
+        if (!hasPrivateKeys(leftPath.node)) return;
+        // for ({ #x: x } of cls);
+        // transforms to:
+        // for (const ref of cls) { ({ #x: x } = ref); body; }
+        // This transform assumes that any expression within the pattern
+        // does not interfere with the iterable `cls`.
+        const temp = scope.generateUidIdentifier("ref");
+        node.left = variableDeclaration("const", [
+          variableDeclarator(temp, null),
+        ]);
+        const assignExpr = expressionStatement(
+          assignmentExpression("=", leftPath.node, cloneNode(temp)),
+        );
+        unshiftForXStatementBody(path, [assignExpr]);
+        scope.crawl();
+      }
+    },
+    VariableDeclaration(path, state) {
+      const { scope, node } = path;
+      const { declarations } = node;
+      if (!declarations.some(declarator => hasPrivateKeys(declarator.id))) {
+        return;
+      }
+      const newDeclarations = [];
+      for (const declarator of declarations) {
+        for (const { left, right } of transformPrivateKeyDestructuring(
+          // @ts-expect-error The id of a variable declarator must not be a RestElement
+          declarator.id,
+          declarator.init,
+          scope,
+          /* isAssignment */ false,
+          /* shouldPreserveCompletion */ false,
+          name => state.addHelper(name),
+          objectRestNoSymbols,
+          /* useBuiltIns */ true,
+        )) {
+          newDeclarations.push(variableDeclarator(left, right));
+        }
+      }
+      node.declarations = newDeclarations;
+      scope.crawl();
+    },
+
+    AssignmentExpression(path, state) {
+      const { node, scope, parent } = path;
+      if (!hasPrivateKeys(node.left)) return;
+      const assignments = [];
+      const shouldPreserveCompletion =
+        (!isExpressionStatement(parent) && !isSequenceExpression(parent)) ||
+        path.isCompletionRecord();
```

Comment:
Why do we exclude `ExpressionStatement` and `SequenceExpression`?

Reply:
It is copied from
https://github.com/babel/babel/blob/9cb5f6c983b6d4be94de880946c16970a3788c9f/packages/babel-plugin-transform-destructuring/src/util.ts#L666

`NodePath#isCompletionRecord` essentially checks alongside the ancestry if the path is the last item of its container. This approach works for ExpressionStatement in Block and expression in SequenceExpression.